tiledb/api/src/array/fragment_info.rs

```diff
@@ -0,0 +1,667 @@
+use std::ops::Deref;
+
+use anyhow::anyhow;
+// use serde::{Deserialize, Serialize};
+// use serde_json::json;
+
+use crate::array::schema::{RawSchema, Schema};
+use crate::config::{Config, RawConfig};
+use crate::context::{CApiInterface, Context, ContextBound};
+use crate::datatype::Datatype;
+use crate::error::Error;
+use crate::fn_typed;
+use crate::key::LookupKey;
+use crate::non_empty_domain::{NonEmptyDomain, Range};
+use crate::string::{RawTDBString, TDBString};
+use crate::Result as TileDBResult;
+
+pub(crate) enum RawFragmentInfo {
+    Owned(*mut ffi::tiledb_fragment_info_t),
+}
+
+impl Deref for RawFragmentInfo {
+    type Target = *mut ffi::tiledb_fragment_info_t;
+    fn deref(&self) -> &Self::Target {
+        let RawFragmentInfo::Owned(ref ffi) = *self;
+        ffi
+    }
+}
+
+impl Drop for RawFragmentInfo {
+    fn drop(&mut self) {
+        let RawFragmentInfo::Owned(ref mut ffi) = *self;
+        unsafe {
+            ffi::tiledb_fragment_info_free(ffi);
+        }
+    }
+}
+
+pub struct FragmentInfo<'ctx> {
+    context: &'ctx Context,
+    raw: RawFragmentInfo,
+}
+
+impl<'ctx> ContextBound<'ctx> for FragmentInfo<'ctx> {
+    fn context(&self) -> &'ctx Context {
+        self.context
+    }
+}
+
+impl<'ctx> FragmentInfo<'ctx> {
```

Comment:
Lots of the methods take a fragment index.
This is really a list of fragments with some extra metadata on top, right?
What do you think about embedding the elements in another structure which binds the index?
And then this parent has `fn fragment(idx: usize) -> TileDBResult<Fragment>`

Reply:
I waffled on this as well. My general preference is to err on the side of transparent wrappers because I've been bitten too many times by wrapper abstractions that prevent legitimate uses of an API. Though I think in this particular case you've convinced me to at least give the convenience approach a whirl.

Reply:
I ended up doing the following:

1. Rename FragmentInfo -> FragmentInfoInternal and made it a private type.
2. FragmentInfo is now a struct holds references to FragmentInfoInternal and an index
3. FragmentInfoList now has the non-index methods and can return specific fragments or iterate over all of them.